### PR TITLE
added registration with email verification

### PR DESCRIPTION
**Pull request by Zomba4okk**

### CODE REVIEW

File: users/models.py [hunk 1]
```diff
@@ -20,6 +20,11 @@ class User(AbstractUser):
         verbose_name='User\'s photo',
         blank=True)
     zip_code = models.CharField(max_length=12, blank=True)
+    is_active = models.BooleanField(
```
Comment (vitali-r) [inline users/models.py]: It's better to add a separate field named "email_confirmed" or smth like that. "is_active" can be used to delete users

File: templates/after_confirmation.html [hunk 2]
```diff
@@ -0,0 +1 @@
+<h2>You are successfullu activated your account. Now you can sign in.</h2>
```
Comment (vitali-r) [inline templates/after_confirmation.html]: You have successfully activated your account.

File: users/serializers.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+UserModel = get_user_model()
+
+
+class UserSerializer(serializers.ModelSerializer):
+    password = serializers.CharField(
+        min_length=8,
+        max_length=50,
+        style={'input_type': 'password'},
+        write_only=True)
+    password2 = serializers.CharField(
```
Comment (vitali-r) [inline users/serializers.py:48]: Why do we need this field?

File: users/serializers.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    class Meta:
+        model = UserModel
+        fields = ('username', 'password', 'password2', 'email')
+
+    def create(self, validated_data):
+        user_object = UserModel(
+            username=validated_data.get('username'),
+            email=validated_data.get('email'))
+        user_object.set_password(validated_data.get('password'))
```
Comment (vitali-r) [inline users/serializers.py]: As password is required it's better to use:
validated_data['password']
the same for the email

File: users/views.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        return HttpResponse(render_to_string('failed_activation.html'))
+
+
+class RegistrationView(APIView):
+    user_model = get_user_model()
+    permission_classes = (permissions.AllowAny,)
+    serializer_class = UserSerializer
+
+    def post(self, request, format=None):
+        data = UserSerializer(data=request.data)
```
Comment (vitali-r) [inline users/views.py]: usually:
serializer = UserSerializer(data=request.data)
(rename variable)

File: users/views.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+            user.save()
+            return HttpResponse(render_to_string('after_confirmation.html'))
+        else:
+            return HttpResponse(render_to_string('failed_activation.html'))
+    except(TypeError, ValueError, OverflowError, user_model.DoesNotExist):
+        return HttpResponse(render_to_string('failed_activation.html'))
+
+
+class RegistrationView(APIView):
+    user_model = get_user_model()
```
Comment (vitali-r) [inline users/views.py]: Is it used anywhere?

File: users/tokens.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+from django.contrib.auth.tokens import PasswordResetTokenGenerator
+from django.utils import six
+
+
+class TokenGenerator(PasswordResetTokenGenerator):
+    def make_hash_value(self, user, timestamp):
```
Comment (dkruk) [inline users/tokens.py]: If I'm not mistaken Token Generators don't use method `make_hash_value`, they are using `_make_hash_value` method.

File: users/views.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    def post(self, request, format=None):
+        serializer = UserSerializer(data=request.data)
+        if serializer.is_valid():
+            user = serializer.save()
+            current_site = get_current_site(request)
+            mail_subject = 'Account activation'
+            message = render_to_string('activation_message.html', {
+                'username': user.username,
+                'domaim': current_site.domain,
```
Comment (dkruk) [inline users/views.py]: in the template expected other variable

File: users/models.py [hunk 9]
```diff
@@ -14,16 +14,20 @@ class User(AbstractUser):
...
     address_room_number = models.CharField(max_length=10, blank=True)
-    phone = models.CharField(max_length=18)
+    phone = models.CharField(max_length=18, blank=True)
     avatar = models.ImageField(
         upload_to='photos/profiles',
         verbose_name='User\'s photo',
         blank=True)
     zip_code = models.CharField(max_length=12, blank=True)
+    email_confirmed = models.BooleanField(default=False)
+    username = models.CharField(max_length=150, verbose_name='username')
```
Comment (dkruk) [inline users/models.py]: Not sure I understand the meaning of this field. Why do we need it?

File: users/models.py [hunk 10]
```diff
@@ -14,16 +14,20 @@ class User(AbstractUser):
...
+    REQUIRED_FIELDS = ['username']
+    USERNAME_FIELD = 'email'
 
     def get_adress(self):
```
Comment (dkruk) [inline users/models.py]: Fix type `address`

File: templates/failed_activation.html [hunk 11]
```diff
@@ -0,0 +1 @@
+<h2>Your link is invalid!</h2>>
```
Comment (dkruk) [inline templates/failed_activation.html]: You don't need this template, return the error message in JSON format

File: templates/after_registration.html [hunk 12]
```diff
@@ -0,0 +1 @@
+<h2>Thank you for choosing our company.<br>Please confirm your email address to complete the registration</h2>
```
Comment (dkruk) [inline templates/after_registration.html]: You don't need this template, return the message in JSON format.

File: templates/after_confirmation.html [hunk 13]
```diff
@@ -0,0 +1 @@
+<h2>You have successfully activated your account. Now you can sign in.</h2>
```
Comment (dkruk) [inline templates/after_confirmation.html]: You don't need this template, return the message in JSON format.

File: users/views.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        user = user_model.objects.get(pk=uid)
+        if account_activation_token.check_token(user, token):
+            user.email_confirmed = True
+            user.save()
+            return HttpResponse(render_to_string('after_confirmation.html'))
+        else:
+            return HttpResponse(render_to_string('failed_activation.html'))
+    except(TypeError, ValueError, OverflowError, user_model.DoesNotExist):
+        return HttpResponse(render_to_string('failed_activation.html'))
+
```
Comment (dkruk) [inline users/views.py:36]: You don't need HTML templates for the response.

**Review by dkruk — CHANGES_REQUESTED** [review]

File: users/models.py (hunk 9 above)
Comment (vitali-r) [reply users/models.py]: @dkrukouski I think the idea is to make the field not unique. But question to @Zomba4ok: why didn't you do the field nullable?

**Review by dkruk — APPROVED** [review]

**Pull request merged**